fastlane/actions/analyze_commits_canary.rb

```diff
@@ -0,0 +1,399 @@
+          # If there is no tag found we taking the first commit of current branch
+          hash = Actions.sh('git rev-list --max-parents=0 HEAD', log: false).chomp
+        else
+          major, minor, patch, preid, pre_ver = break_apart_tag(tag: tag).values
+
+          tag_name = "#{major}.#{minor}.#{patch}"
+
+          if !preid.nil?
+            tag_name += "-#{preid}.#{pre_ver}"
+          end
+
+          parsed_version = tag_name.match(params[:tag_version_match])
+
+          if parsed_version.nil?
+            UI.user_error!("Error while parsing version from tag #{tag_name} by using tag_version_match - #{params[:tag_version_match]}. Please check if the tag contains version as you expect and if you are using single brackets for tag_version_match parameter.")
+          end
+
+          # Get a hash of last version tag
+          hash = get_last_tag_hash(tag_name: tag_name)
+
+          UI.message("Found a tag #{tag} associated with version #{tag_name}")
+        end
+
+        version = tag_name
+
+        next_major = major
+        next_minor = minor
+        next_patch = patch
+
+        if preid.nil?
+          next_patch += 1
+          next_preid = params[:preid]
+          next_pre_ver = 0
+        else
+          # different preid exists on tag than the one that was passed in
+          if preid != params[:preid]
+            next_preid = params[:preid]
+            next_pre_ver = 0
+          else
+            # if preid exists and matches the one that was passed in, increment the prerelease version
+            next_preid = preid
+            next_pre_ver = pre_ver + 1
+          end
+        end
+
+        next_version = "#{next_major}.#{next_minor}.#{next_patch}-#{next_preid}.#{next_pre_ver}"
+
+        is_next_version_releasable = true
+
+        Actions.lane_context[SharedValues::RELEASE_ANALYZED] = true
+        Actions.lane_context[SharedValues::RELEASE_IS_NEXT_VERSION_HIGHER] = is_next_version_releasable
+        # Last release analysis
+        Actions.lane_context[SharedValues::RELEASE_LAST_TAG_HASH] = hash
+        Actions.lane_context[SharedValues::RELEASE_LAST_VERSION] = version
+        # Next release analysis
+        Actions.lane_context[SharedValues::RELEASE_NEXT_MAJOR_VERSION] = next_major
+        Actions.lane_context[SharedValues::RELEASE_NEXT_MINOR_VERSION] = next_minor
+        Actions.lane_context[SharedValues::RELEASE_NEXT_PATCH_VERSION] = next_patch
+        Actions.lane_context[SharedValues::RELEASE_NEXT_VERSION] = next_version
+        Actions.lane_context[SharedValues::RELEASE_NEXT_PRERELEASE_VERSION] = "#{next_preid}.#{next_pre_ver}"
+
+        success_message = "Next version (#{next_version}) is higher than last version (#{version}). This version should be released."
+        UI.success(success_message) if is_next_version_releasable
+
+        is_next_version_releasable
+      end
+
+      def self.is_codepush_friendly(params)
```

Comment:
As above, this reads like a simple evaluation, but you're also updating `lane_context`. Consider breaking the side effects out